mijia/src/decode.rs

```diff
@@ -0,0 +1,83 @@
+use std::cmp::max;
+use std::convert::TryInto;
+use std::fmt::{Display, Formatter};
+
+/// A set of readings from a Mijia sensor.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Readings {
+    /// Temperature in ºC, with 2 decimal places of precision
+    pub temperature: f32,
+    /// Percent humidity
+    pub humidity: u8,
+    /// Voltage in millivolts
+    pub battery_voltage: u16,
+    /// Inferred from `battery_voltage` with a bit of hand-waving.
+    pub battery_percent: u16,
+}
+
+impl Display for Readings {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "Temperature: {:.2}ºC Humidity: {:?}% Battery: {:?} mV ({:?}%)",
+            self.temperature, self.humidity, self.battery_voltage, self.battery_percent
+        )
+    }
+}
+
+impl Readings {
+    /// Decode the readings from the raw bytes of the Bluetooth characteristic value, if they are
+    /// valid.
+    /// Returns `None` if the value is not valid.
+    pub fn decode(value: &[u8]) -> Option<Readings> {
```

Comment:
long-term, I would be tempted to implement https://doc.rust-lang.org/std/convert/trait.TryFrom.html , and then use `let readings = value.try_into().ok()?` in the one place where it's used.

For now, I don't think that consumers will ever have `&[u8]` in their hand, so maybe `pub(crate)` and leave everything else as-is?

Reply:
Done the latter.